prometheus-to-sd/main.go

```diff
@@ -138,40 +135,64 @@ func readAndPushDataToStackdriver(stackdriverService *v3.Service, gceConf *confi
+			metricDescriptorCache.MarkStale()
+			glog.V(4).Infof("Scraping metrics of component %v", sourceConfig.Component)
+			select {
+			case <-signal:
+				glog.V(4).Infof("Updating metrics cache for component %v", sourceConfig.Component)
+				metricDescriptorCache.Refresh()
+				if useWhitelistedMetricsAutodiscovery {
+					sourceConfig.UpdateWhitelistedMetrics(metricDescriptorCache.GetMetricNames())
+					glog.V(2).Infof("Autodiscovered whitelisted metrics for component %v: %v", commonConfig.ComponentName, sourceConfig.Whitelisted)
+				}
+				signal = time.After(*metricDescriptorsResolution)
+			default:
 			}
-			signal = time.After(*metricDescriptorsResolution)
-		default:
-		}
-		if useWhitelistedMetricsAutodiscovery && len(sourceConfig.Whitelisted) == 0 {
-			glog.V(4).Infof("Skipping %v component as there are no metric to expose.", sourceConfig.Component)
-			continue
-		}
-		metrics, err := translator.GetPrometheusMetrics(&sourceConfig)
-		if err != nil {
-			glog.V(2).Infof("Error while getting Prometheus metrics %v for component %v", err, sourceConfig.Component)
-			continue
-		}
-		if *omitComponentName {
-			metrics = translator.OmitComponentName(metrics, sourceConfig.Component)
+			if useWhitelistedMetricsAutodiscovery && len(sourceConfig.Whitelisted) == 0 {
+				glog.V(4).Infof("Skipping %v component as there are no metric to expose.", sourceConfig.Component)
+				continue
+			}
+			metrics, err := translator.GetPrometheusMetrics(&sourceConfig)
+			if err != nil {
+				glog.V(2).Infof("Error while getting Prometheus metrics %v for component %v", err, sourceConfig.Component)
+				continue
+			}
+			if *omitComponentName {
+				metrics = translator.OmitComponentName(metrics, sourceConfig.Component)
+			}
+			if strings.HasPrefix(commonConfig.GceConfig.MetricsPrefix, customMetricsPrefix) {
+				metricDescriptorCache.UpdateMetricDescriptors(metrics, sourceConfig.Whitelisted)
+			} else {
+				metricDescriptorCache.ValidateMetricDescriptors(metrics, sourceConfig.Whitelisted)
+			}
+			ts := translator.TranslatePrometheusToStackdriver(commonConfig, sourceConfig.Whitelisted, metrics, metricDescriptorCache)
+			translator.SendToStackdriver(stackdriverService, commonConfig, ts)
 		}
-		if strings.HasPrefix(commonConfig.GceConfig.MetricsPrefix, customMetricsPrefix) {
-			metricDescriptorCache.UpdateMetricDescriptors(metrics, sourceConfig.Whitelisted)
-		} else {
-			metricDescriptorCache.ValidateMetricDescriptors(metrics, sourceConfig.Whitelisted)
+	}
+}
+
+func handleDynamicConfigurations(stackdriverService *v3.Service, gceConf *config.GceConfig, sourceConfigChangeCh <-chan config.SourceConfigChange) {
+	channels := make(map[string]chan bool)
+	for sourceConfigChange := range sourceConfigChangeCh {
+		switch sourceConfigChange.ChangeType {
+		case config.Added, config.Modified:
+			if sourceConfigChange.SourceConfig.Host != "" && channels[sourceConfigChange.SourceConfig.Component] == nil {
```

Comment:
What checking not empty host is used for? are you expecting it? in what circumstances?

Reply:
We need the host/ip to be able to make requests to the component, when a pod is firstly created it does not have an ip yet